### PR TITLE
Add support for placeholders

### PR DESCRIPTION
**Pull request by kandl**
Placeholder represents a reference to a specific part of the execution - attribute(s), measure(s), filter(s), sort(s) or total(s),
that may change the value at runtime. You can provide them to visualizations instead of the attributes/measures/filters/sorts/totals themselves.
Placeholders will be replaced with the actual values during their resolution.

Placeholders are built on top of React context and hooks.

- Implement placeholders context & hooks
- Add placeholders support to all visualizations & relevant execution components/hooks
- Add docs
- Add examples

JIRA: RAIL-2800, RAIL-3045

---

Supported PR commands:

| Command                | Description            |
| ---------------------- | ---------------------- |
| `ok to test`           | Re-run standard checks |
| `extended test`        | BackstopJS tests       |
| `extended check sonar` | SonarQube tests        |

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `check-extended` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/3717/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/5339/ : FAILURE in 12m 31s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/3718/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/5340/ : FAILURE in 12m 36s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/3721/ : SUCCESS in 14s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/5344/ : SUCCESS in 12m 49s

File: libs/sdk-ui-charts/api/sdk-ui-charts.api.md [hunk 1]
```diff
@@ -23,16 +22,18 @@ import { ITheme } from '@gooddata/sdk-backend-spi';
 import { IVisualizationCallbacks } from '@gooddata/sdk-ui';
 import { IVisualizationProps } from '@gooddata/sdk-ui';
 import { default as React_2 } from 'react';
+import { ValueOrPlaceholder } from '@gooddata/sdk-ui';
+import { ValuesOrPlaceholders } from '@gooddata/sdk-ui';
 import { VisType } from '@gooddata/sdk-ui';
 
 // @public
-export const AreaChart: import("react").ComponentType<IAreaChartProps>;
+export const AreaChart: (props: IAreaChartProps) => JSX.Element;
```
Comment (lupko) [inline libs/sdk-ui-charts/api/sdk-ui-charts.api.md:30]: just doublechecking that this change is compatible?

File: libs/sdk-ui-charts/api/sdk-ui-charts.api.md [hunk 2]
```diff
@@ -130,12 +133,13 @@ export interface IBaseChartProps extends ICoreChartProps {
...
-    sortBy?: ISortItem[];
-    viewBy?: IAttribute;
-    xAxisMeasure?: IMeasure;
-    yAxisMeasure?: IMeasure;
+    filters?: ValuesOrPlaceholders<INullableFilter>;
+    placeholdersResolutionContext?: any;
+    size?: ValueOrPlaceholder<AnyMeasure>;
+    sortBy?: ValuesOrPlaceholders<ISortItem>;
+    viewBy?: ValueOrPlaceholder<IAttribute>;
+    xAxisMeasure?: ValueOrPlaceholder<AnyMeasure>;
```
Comment (lupko) [inline libs/sdk-ui-charts/api/sdk-ui-charts.api.md:141]: isn't this widening the type? before we have IMeasure only. with AnyMeasure, user can add IMeasure or IAttribute or their placeholders right?

File: libs/sdk-ui-geo/api/sdk-ui-geo.api.md [hunk 3]
```diff
@@ -36,7 +38,7 @@ export type CenterPositionChangedCallback = (center: IGeoLngLat) => void;
 export const CoreGeoChart: React_2.FC<ICoreGeoChartProps & WrappedComponentProps>;
 
 // @public (undocumented)
-export const GeoPushpinChart: React_2.ComponentType<Pick<IGeoPushpinChartProps, "locale" | "color" | "size" | "location" | "onError" | "config" | "onDrill" | "drillableItems" | "afterRender" | "onCenterPositionChanged" | "onZoomChanged" | "exportTitle" | "ErrorComponent" | "LoadingComponent" | "onExportReady" | "onLoadingChanged" | "pushData" | "backend" | "workspace" | "segmentBy" | "filters" | "sortBy">>;
+export const GeoPushpinChart: (props: IGeoPushpinChartProps) => JSX.Element;
```
Comment (lupko) [inline libs/sdk-ui-geo/api/sdk-ui-geo.api.md:41]: why this?

File: libs/sdk-ui-charts/api/sdk-ui-charts.api.md (hunk 1 above)
Comment (kandl) [reply libs/sdk-ui-charts/api/sdk-ui-charts.api.md:30]: Actually, it's a "breaking" change (React.ComponentType accepts also "children" prop, React.ComponentType could represent also class component), but I guess this signature without the children prop is more correct - our charts shouldn't accept children prop (it does nothing).

File: libs/sdk-ui/src/base/react/placeholders/tests/useComposedPlaceholder.test.tsx [hunk 4]
```diff
@@ -0,0 +1,103 @@
+// (C) 2019-2021 GoodData Corporation
+import React from "react";
+import { newMeasure } from "@gooddata/sdk-model";
+import { mount } from "enzyme";
+import { PlaceholdersProvider, IPlaceholdersProviderProps } from "../context";
+import { newComposedPlaceholder, newPlaceholder } from "../factory";
+import { IComposedPlaceholder } from "../base";
+import { useComposedPlaceholder } from "../hooks";
+import { newRelativeDateFilter } from "../../../../../../sdk-model/src/execution/filter/factory";
+import { modifySimpleMeasure } from "../../../../../../sdk-model/src/execution/measure/factory";
```
Comment (lupko) [inline libs/sdk-ui/src/base/react/placeholders/tests/useComposedPlaceholder.test.tsx:10]: bad imports

File: libs/sdk-ui/src/execution/useExecution.ts [hunk 5]
```diff
@@ -12,27 +18,32 @@ export interface IUseExecutionConfig {
...
     /**
      * Optional sorting to apply on server side.
      */
-    sortBy?: ISortItem[];
+    sortBy?: ValuesOrPlaceholders<ISortItem>;
+
+    /**
+     * Optional resolution context for composed placeholders.
+     */
+    placeholdersResolutionContext?: any;
```
Comment (lupko) [inline libs/sdk-ui/src/execution/useExecution.ts:46]: these definitions are copy-pasted in multiple places. no need to fix now, it was existing problem. please create followup under FET-512 to deduplicate.

File: libs/sdk-ui-charts/api/sdk-ui-charts.api.md (hunk 1 above)
Comment (lupko) [reply libs/sdk-ui-charts/api/sdk-ui-charts.api.md:30]: ok, seems more like a bugfix then. let's keep it.

File: libs/sdk-ui-geo/api/sdk-ui-geo.api.md (hunk 3 above)
Comment (kandl) [reply libs/sdk-ui-geo/api/sdk-ui-geo.api.md:41]: Before:
`
const GeoPushpinChart = withTheme(withContexts(GeoPushpinChartInner));
`
Now:
`
const WrappedGeoPushpinChart = withTheme(withContexts(GeoPushpinChartInner));

/**
 * @public
 */
export const GeoPushpinChart = (props: IGeoPushpinChartProps) => {
    const [location, size, color, segmentBy, filters, sortBy] = useResolveValuesWithPlaceholders(
        [props.location, props.size, props.color, props.segmentBy, props.filters, props.sortBy],
        props.placeholdersResolutionContext,
    );

    return <WrappedGeoPushpinChart {...props} {...{ location, size, color, segmentBy, filters, sortBy }} />;
};
`

File: libs/sdk-ui/src/base/react/placeholders/tests/useComposedPlaceholder.test.tsx (hunk 4 above)
Comment (kandl) [reply libs/sdk-ui/src/base/react/placeholders/tests/useComposedPlaceholder.test.tsx]: Corrected

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/3728/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/5350/ : SUCCESS in 12m 44s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/3729/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/5351/ : SUCCESS in 12m 38s

**Comment by lupko** [comment]
extended test

**Comment by yenkins** [comment]
Link to Storybook test result: https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/941/artifact/libs/sdk-ui-tests/backstop/output/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/3733/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/941/ : SUCCESS in 24m 29s

**Review by lupko — APPROVED** [review]
LGTM

**Comment by yenkins** [comment]
### Sonar scan result
 - No issue found. Congratulations!

More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR1374

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Pull request merged**